Add support for boxed pointers

densh opened the pull request:
This PR introduces two major changes:

### Makes `Ptr[T]` a managed object

So now pointers are:

* Can be used with `asInstanceOf`, so there is no more need for our own special `a.cast[T]` syntax.
* Be passed to generic context (i.e., generic methods that expect boxed representation).
* Have meaningful `toString`, `hashCode` and `equals` based on the value of underlying pointer.
* Are stored in arrays and Scala collections in boxed form. 

Moreover:

* All pointer operations are now well-behaved in generic context as long as there is a provided instance of `native.Tag`. This makes it possible to write type-polymorphic code over the contents of the pointer itself.

### Makes `CArray` a managed object

* Well behaved with respect to boxing.
* Has meaningful `toString`, `hashCode` and `equals` based on the value of underlying pointer.
* Can access elements directly without going back to pointers.

Additionally, we also change encoding for `Nat`. Intead of `Nat.Digit[A, B]` we have multiple overloaded `Nat.DigitN` that correspond to number of digits in the number. So for example `128` is represented as `Nat.Digit3[Nat._1, Nat._2, Nat._8]`.

### Makes `CStruct` a managed object

* Well behaved with respect to boxing.
* Has meaningful `toString`, `hashCode` and `equals` based on the value of underlying pointer.
* Can access elements directly without going back to pointers.

### Changes encoding of function pointers to new SAM-based `CFuncPtrN`.

Similarly to normal pointers:

* Function pointers are now regular objects without any special semantics.
* Declaring a function pointer auto-generates an extern-accessible bridge method.
* Passing a function pointer to extern methods resolves a bridge method and passes that as a pointer value.

Note this still doesn't allow function pointers to have any state. 

Fixes #367 #396

densh commented:
This pr depends on #1411 for performance reasons. Current implementation is semantically correct but introduces a major performance regression until we get inlining merged.

densh commented on nativelib/src/main/scala/scala/scalanative/native/Ptr.scala:
```diff
@@ -1,142 +1,187 @@
 
-/** The C `const T *` pointer. */
-final abstract class Ptr[T] {
+final class Ptr[T](private[scalanative] val rawptr: RawPtr) {
```
@shadaj Here is a concrete example of what I meant in the discussion of #1420. `Ptr` type is a wrapper over `RawPtr`. User-facing API is safe from the boxing point of view and encapsulates raw unmanaged `RawPtr`. 

Similarly for `Word` boxing to work, `Word` itself could be a wrapper over `RawWord` which maps to a numeric type under the hood:
```
// .../native/Word.scala
final class Word(private[scalanative] val rawword: RawWord) {
  ...
  @inline def +(other: Word): Word =
     new Word(Intrinsics.addWord(this.rawword, other.rawword))
  ...
}

// .../runtime/RawWord.scala
final abstract class RawWord

// .../runtime/Intrinsics.scala
object Intrinsics {
  ...
  def addWord(l: RawWord, r: RawWord): RawWord = intrinsic
  ...
}
```

shadaj replied:
Awesome, thanks for the example! The semester just started so still adjusting to the workload in my new schedule, but I'll try to update my PR to use this style sometime this week.

densh commented on unit-tests/src/test/scala/scala/scalanative/native/SizeofSuite.scala:
```diff
@@ -75,9 +75,9 @@ object SizeofSuite extends tests.Suite {
-  type _1024 = Nat.Digit[Nat._1, Nat.Digit[Nat._0, Nat.Digit[Nat._2, Nat._4]]]
+  type _32   = Nat.Digit2[Nat._3, Nat._2]
+  type _128  = Nat.Digit3[Nat._1, Nat._2, Nat._8]
+  type _1024 = Nat.Digit4[Nat._1, Nat._0, Nat._2, Nat._4]
```
/cc @rwhaling This is the new Nat encoding that I mentioned at some point. IMO it looks nicer actually. 😄

densh replied:
With this change, sizeof is always constant folded away and is zero cost despite being completely implemented in the library code.

densh commented:
Recent changes also make `CArray` and `CStructN` behave like regular objects, similarly to `Ptr`.

ekrich commented:
This is going to make a huge improvement to making coding easier. 👍

densh commented:
There are no major performance regressions after this change on our benchmarks. Typical uses of pointers are optimized away and boxing doesn't cause any major overhead. This is quite amazing given how much indirection is there in the current implementation (we removed special-case code from the nscplugin and moved it to the library so optimizer has more work to do).

densh commented:
@ekrich I'm currently experimenting with improving usability of structs. In current prototype `Ptr -> CStructN` is a no-op conversion that can be done implicitly. I'm not sure if I'm going to keep yet though.

ekrich commented:
@densh Not explicitly de-referencing `Ptr` looks nice in code but I can understand if you didn't want to keep it. Does the pass by value for `struct` look attainable?

densh commented:
Delayed structs seem to be really tricky to get right so I'll do it in a separate PR.

densh commented:
@LeeTibbert It looks like unit testing overflow bug has hit me this time. Time to get this over with, it's caused enough pain already.

densh commented:
Current if you mix boxing, c-style varargs and new revised structs in this PR:
```
import scalanative.native._
import scalanative.libc._

object Main {
  type Point = CStruct2[CInt,CInt]
  def main(args:Array[String]):Unit = {
    val p:Ptr[Byte] = stdlib.malloc(sizeof[Point])
    val point:Ptr[Point] = p.asInstanceOf[Ptr[Point]]
    point._1 = 5
    point._2 = 8
    stdio.printf(c"(printf) point at %p: (%d,%d)\n",point,point._1,point._2)
    println(s"(println) point: $point, x: ${point._1}, y: ${point._2}")
  }
}
```
It doesn't do what you expect and prints wrong value for the printf call. As discovered by @rwhaling.

densh commented:
So thanks to the nudge from @rwhaling we now have 100% generic and abstraction-friendly `CVarArg` and `CVarArgList`. This fixes #153, and #1142.

densh commented:
@ekrich Currently we only have one-way interop: Scala can call C. If/when we make it work the other way around (i.e. #155) we might add support for reading values from `CVarArgList` as well.

densh commented on nativelib/src/main/scala/scala/scalanative/native/CVarArgList.scala:
```diff
@@ -0,0 +1,110 @@
+object CVarArgList {
+  // Implementation below is based on VarArgs.swift from apple/swift repo.
+  // Currently we only support X86_64, we'll need to revisit the code below
+  // if and when we add any more officially supported architectures.
```
/cc @shadaj
